module/zfs/zvol.c

```diff
@@ -77,11 +96,140 @@ typedef struct zvol_state {
 	dev_t			zv_dev;		/* device id */
 	struct gendisk		*zv_disk;	/* generic disk */
 	struct request_queue	*zv_queue;	/* request queue */
+	zvol_handle_t		zv_handle;	/* unique key for the state */
+	refcount_t		zv_use_count;	/* keeps the state in memory */
 	list_node_t		zv_next;	/* next zvol_state_t linkage */
 } zvol_state_t;
 
+typedef enum {
+	ZVOL_ASYNC_CREATE_MINORS,
+	ZVOL_ASYNC_REMOVE_MINORS,
+	ZVOL_ASYNC_RENAME_MINORS,
+	ZVOL_ASYNC_MAX
+} zvol_async_op_t;
+
+typedef struct {
+	zvol_async_op_t op;
+	char pool[MAXNAMELEN];
+	char name1[MAXNAMELEN];
+	char name2[MAXNAMELEN];
+} zvol_task_t;
+
 #define	ZVOL_RDONLY	0x1
 
+static int
+zvol_handle_compare(const void *n1, const void *n2) {
+	zvol_handle_t h1 = ((zvol_handle_node_t *)n1)->zhn_handle;
+	zvol_handle_t h2 = ((zvol_handle_node_t *)n2)->zhn_handle;
+
+	if (h1 > h2)
+		return (1);
+	else if (h1 < h2)
+		return (-1);
+	else
+		return (0);
+}
+
+static void
+zvol_handle_init(void)
+{
+	avl_create(&zvol_handle_tree, zvol_handle_compare,
+	    sizeof (zvol_handle_node_t),
+		offsetof(struct	zvol_handle_node, zhn_node));
+}
+
+static void
+zvol_handle_destroy(void)
+{
+	zvol_handle_node_t *node;
+	void *cookie = NULL;
+
+	while ((node = avl_destroy_nodes(&zvol_handle_tree, &cookie)) != NULL)
+		kmem_free(node, sizeof (zvol_handle_node_t));
+	avl_destroy(&zvol_handle_tree);
+}
+
+static zvol_state_t *
+zvol_handle_hold(void *private_data)
+{
+	zvol_handle_t handle = (zvol_handle_t)private_data;
+	zvol_handle_node_t what, *node;
+	boolean_t drop = B_FALSE;
+
+	what.zhn_handle = handle;
+
+	if (!MUTEX_HELD(&zvol_state_lock)) {
+		mutex_enter(&zvol_state_lock);
+		drop = B_TRUE;
+	}
+
+	node = avl_find(&zvol_handle_tree, &what, NULL);
+	if (node)
+		refcount_add(&node->zhn_state->zv_use_count, zvol_tag);
+
+	if (drop)
+		mutex_exit(&zvol_state_lock);
+
+	return (node) ? (node->zhn_state) : (NULL);
+}
+
+static void
+zvol_handle_rele(zvol_state_t *state)
+{
+	boolean_t drop = B_FALSE;
+
+	if (!MUTEX_HELD(&zvol_state_lock)) {
+		mutex_enter(&zvol_state_lock);
+		drop = B_TRUE;
+	}
+
+	refcount_remove(&state->zv_use_count, zvol_tag);
+
+	if (drop)
+		mutex_exit(&zvol_state_lock);
+}
+
+static void *
+zvol_handle_register(zvol_state_t *state)
+{
+	zvol_handle_node_t what, *node;
+	avl_index_t where;
+
+	ASSERT(MUTEX_HELD(&zvol_state_lock));
+	/* Valid handles start with 1 */
+	what.zhn_handle = ++zvol_handles;
```

Comment:
While this should never wraparound in practice, it makes me a little nervous. We should at least have a comment explaining that we assume that we will never wraparound because doing 1 increment a nanosecond (which isn't possible on any current or future hardware) would take more than 584 years to wraparound. Maybe something to handle a potentially weird state of wrapping around would be helpful if only to make the danger clear to anyone modifying this in the future.